File: src/typegraph/rust/typegraph/src/runtimes/random.rs
// Copyright Metatype OÜ, licensed under the Elastic License 2.0.
// SPDX-License-Identifier: Elastic-2.0

use crate::{
    t::{self, TypeBuilder, TypeDef},
    wasm::{
        self,
        core::RuntimeId,
        runtimes::{BaseMaterializer, Effect, MaterializerRandom, RandomRuntimeData},
    },
    Result,
};

#[derive(Debug)]
pub struct RandomRuntime {
    id: RuntimeId,
}

impl RandomRuntime {
    pub fn new(seed: Option<u32>, reset: Option<&str>) -> Result<Self> {
        let data = RandomRuntimeData {
            seed,
            reset: reset.map(|r| r.to_string()),
        };

        let id = wasm::with_runtimes(|r, s| r.call_register_random_runtime(s, &data))?;

        Ok(RandomRuntime { id })
    }

    pub fn gen<T: TypeBuilder>(&self, out: T) -> Result<TypeDef> {
        let base = BaseMaterializer {
            runtime: self.id,
            effect: Effect::Read,
        };
        let mat = MaterializerRandom { runtime: self.id };
        let mat_id = wasm::with_runtimes(|r, s| r.call_create_random_mat(s, base, mat))?;

        t::funcb(t::r#struct(), out, mat_id)
    }
}

File: src/typegraph/python/typegraph/runtimes/random.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0
from dataclasses import dataclass
from typing import Optional

from typegraph import t
from typegraph.gen.runtimes import (
    BaseMaterializer,
    MaterializerRandom,
    RandomRuntimeData,
)
from typegraph.runtimes.base import Materializer, Runtime
from typegraph.sdk import runtimes


class RandomRuntime(Runtime):
    def __init__(self, seed: Optional[int] = None, reset: Optional[str] = ""):
        super().__init__(
            runtimes.register_random_runtime(
                data=RandomRuntimeData(seed=seed, reset=reset)
            )
        )

    def gen(
        self,
        out: "t.typedef",
    ):
        effect = "read"

        mat_id = runtimes.create_random_mat(
            base=BaseMaterializer(runtime=self.id, effect=effect),
            data=MaterializerRandom(runtime=self.id),
        )

        return t.func(
            t.struct({}),
            out,
            RandomMat(id=mat_id, runtime=self.id, effect=effect),
        )


@dataclass
class RandomMat(Materializer):
    runtime: int
